safehttp/plugins/xsrf/xsrf.go

```diff
@@ -18,146 +18,229 @@
-			// present.
-			mf, err := r.MultipartForm(32 << 20)
-			if err != nil {
-				return w.WriteError(safehttp.StatusBadRequest)
-			}
-			f = &mf.Form
+			code = safehttp.StatusBadRequest
+			return
 		}
+		f = &mf.Form
+	}
+
+	token = f.String(c.tokenKey, "")
+	if f.Err() != nil || token == "" {
+		code = safehttp.StatusUnauthorized
+		return
+	}
 
-		tok := f.String(TokenKey, "")
-		if f.Err() != nil || tok == "" {
-			return w.WriteError(safehttp.StatusUnauthorized)
+	actionID = r.URL.Path()
+	return
+}
+
+func (c defaultChecker) Validate(token, userID, actionID string) safehttp.StatusCode {
+	if !xsrftoken.Valid(token, c.secretAppKey, userID, actionID) {
+		return safehttp.StatusForbidden
+	}
+	return safehttp.StatusOK
+}
+
+type defaultInjector struct {
+	cookieIDKey  string
+	secretAppKey string
+}
+
+func (i defaultInjector) Inject(resp safehttp.Response, w *safehttp.ResponseWriter, r *safehttp.IncomingRequest) error {
+	c, err := r.Cookie(i.cookieIDKey)
+	if err != nil {
+		buf := make([]byte, 20)
+		if _, err := rand.Read(buf); err != nil {
+			return fmt.Errorf("crypto/rand.Read: %v", err)
 		}
+		c = safehttp.NewCookie(i.cookieIDKey, base64.StdEncoding.EncodeToString(buf))
+		c.SetSameSite(safehttp.SameSiteStrictMode)
 
-		if ok := xsrftoken.Valid(tok, it.SecretAppKey, cookieID.Value(), actionID); !ok {
-			return w.WriteError(safehttp.StatusForbidden)
+		if err := w.SetCookie(c); err != nil {
+			return err
 		}
 	}
 
-	tok := xsrftoken.Generate(it.SecretAppKey, cookieID.Value(), actionID)
-	r.SetContext(context.WithValue(r.Context(), tokenCtxKey{}, tok))
-	return safehttp.NotWritten()
-}
+	tok := xsrftoken.Generate(i.secretAppKey, c.Value(), r.URL.Path())
 
-// Commit adds the XSRF token corresponding to the safehttp.TemplateResponse
-// with key "XSRFToken". The token corresponds to the user information found in
-// the request.
-func (it *Interceptor) Commit(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, resp safehttp.Response, _ safehttp.InterceptorConfig) safehttp.Result {
 	tmplResp, ok := resp.(safehttp.TemplateResponse)
 	if !ok {
-		return safehttp.NotWritten()
-	}
-
-	tok, err := Token(r)
-	if err != nil {
-		// The token should have been added in the Before stage and if that is
-		// not the case, a server misconfiguration occured.
-		return w.WriteError(safehttp.StatusInternalServerError)
+		return nil
 	}
 
 	// TODO(maramihali@): Change the key when function names are exported by
 	// htmlinject
 	// TODO: what should happen if the XSRFToken key is not present in the
 	// tr.FuncMap?
 	tmplResp.FuncMap["XSRFToken"] = func() string { return tok }
-	return safehttp.NotWritten()
+	return nil
+}
+
+type angularChecker struct {
+	tokenCookieName string
+	tokenHeaderName string
+}
+
+func (c angularChecker) Retrieve(r *safehttp.IncomingRequest) (token, userID, actionID string, code safehttp.StatusCode) {
+	cookie, err := r.Cookie(c.tokenCookieName)
+	if err != nil {
+		code = safehttp.StatusForbidden
+		return
```

Comment:
Try to avoid using naked returns, especially if you have patterns like 

```
assignment to return value
naked return
```

This line would be more readable like 
```return "","","", safehttp.StatusForbidden```
since it would highlight what you are doing.

I understand it is harder to write code like this, but I think the readability improvement is worth.

Same for other naked returns.